### ISSUE TITLE
Command line overrides don't take priority over optional overrides in the defaults list

### ISSUE THREAD

**Issue by jbaczek** [facebookresearch/hydra#1971]
# 🐛 Bug
## Description
I want to use a command line package override but order of resolving overrides prevents me from doing so. There is a configuration of model and dataset for which I want a trainer to have specific parameters and by default I want to train with a specific criterion. There is a case in which I want to override this default criterion by importing a different config from the same config group. I can do it in some cases but not in the others. I use yaml configs.

## Checklist
- [ x] I checked on the latest version of Hydra
- [ x] I created a minimal repro (See [this](https://stackoverflow.com/help/minimal-reproducible-example) for tips).

## To reproduce
Config stricture:
```
.
├── conf
│   ├── config.yaml
│   ├── dataset
│   │   ├── DatasetA.yaml
│   │   └── DatasetB.yaml
│   ├── model
│   │   ├── ModelA.yaml
│   │   └── ModelB.yaml
│   ├── model_dataset
│   │   └── ModelA_DatasetA.yaml
│   └── trainer
│       ├── TrainerA.yaml
│       ├── TrainerB.yaml
│       └── criterion
│           ├── CriterionA.yaml
│           └── CriterionB.yaml
└── my_app.py
```
Contents of the files:
```
root@jbaczek-ws:/workspace/tmp# for F in `find . -name "*.yaml"`; do echo $F && cat $F; done
./conf/model/ModelB.yaml
param: B
./conf/model/ModelA.yaml
param: A
./conf/config.yaml
defaults:
  - trainer: TrainerA
  - model: ModelA
  - dataset: DatasetA
  - optional model_dataset@_global_: ${model}_${dataset}
./conf/dataset/DatasetB.yaml
ds_param: 'xxx'
./conf/dataset/DatasetA.yaml
ds_param: 'aaa'
./conf/trainer/criterion/CriterionB.yaml
crit_param1: 888
crit_param2: 999
./conf/trainer/criterion/CriterionA.yaml
crit_param1: 111
crit_param2: 222
./conf/trainer/TrainerB.yaml
defaults:
        - criterion: CriterionB

train_param: 90
./conf/trainer/TrainerA.yaml
defaults:
        - criterion: CriterionA

train_param: 13
./conf/model_dataset/ModelA_DatasetA.yaml
trainer:
        criterion:
                crit_param1: 1337


root@jbaczek-ws:/workspace/tmp# cat my_app.py 
import hydra
@hydra.main(config_path="conf", config_name="config")
def main(config):
    print(config)

if __name__ == "__main__":
    main()

```

** Minimal Code/Config snippet to reproduce **
```
python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionB
python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionA
python my_app.py -c job model=ModelB dataset=DatasetA trainer/criterion=CriterionB
```

** Stack trace/error message **
```
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionA
trainer:
  criterion:
    crit_param1: 1337
    crit_param2: 222
  train_param: 13
model:
  param: A
dataset:
  ds_param: aaa
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionB
trainer:
  criterion:
    crit_param1: 1337
    crit_param2: 999
  train_param: 13
model:
  param: A
dataset:
  ds_param: aaa
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelB dataset=DatasetA trainer/criterion=CriterionB
trainer:
  criterion:
    crit_param1: 888
    crit_param2: 999
  train_param: 13
model:
  param: B
dataset:
  ds_param: aaa
```

## Expected Behavior
Command line package overrides should be resolved after all other overrides. Expected output:
```
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionB
trainer:
  criterion:
    crit_param1: 888
    crit_param2: 999
  train_param: 13
model:
  param: A
dataset:
  ds_param: aaa

```
## System information
- **Hydra Version** :  1.1.1
- **Python version** : 3.8.12
- **Virtual environment type and version** : Docker container based on Nvidia's NGC PyTorch container. `docker pull nvcr.io/nvidia/pytorch:21.12-py3`
- **Operating system** : Ubuntu 20.04

**Comment by Jasha10** [comment]
Hi @jbaczek, thanks for the detailed report! I believe this is intended behavior.

# TLDR
To get the behavior you expect, you need to reorder the defaults list in your `config.yaml` file so that `trainer` comes *after* `model_dataset`:
```yaml
defaults:
  - model: ModelA
  - dataset: DatasetA
  - optional model_dataset@_global_: ${model}_${dataset}
  - trainer: TrainerA
```

# Recursive Defaults Lists in Detail

To explain why we're seeing the behavior you've highlighted, let me write down my understanding of Hydra's defaults-list implementation. I will focus on `ModelA`/`DatasetA`/`CriterionB` case that you highlighted in the "Expected Behavior" section of your comment above:
```text
python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionB
```

First, a high-level summary of how Hydra resolves defaults and overrides:

To construct the output config (which will be passed to your app), Hydra needs to figure out which input configs (e.g. yaml files or structured configs) should be merged together. To do this, hydra first creates a "defaults tree" which represents the hierarchy of input configs. Command-line overrides are accounted for in construction of this tree. This tree of defaults is then flattened into a list by traversing the nodes of the tree depth-first, with child nodes being listed before their parents. Finally, the configs in that list are merged, with later configs overwriting earlier ones.

Now, let's dive into the details.

### The list of overrides
Hydra maintains a list of "overrides". This list is compiled by combining overrides passed at the command line with overrides specified in defaults lists via the [`override`](https://hydra.cc/docs/next/advanced/defaults_list/#introduction) keyword. I believe that the CLI overrides take precedent over the override-keyword overrides.
In the given A/A/B case, the list of overrides should look like this:
```
* model=ModelA
* dataset=DatasetA
* trainer/criterion=CriterionB
```


### Recursive construction of the defaults tree
Starting with your main config, `config.yaml`, Hydra creates a defaults tree. Creating the tree is a recursive procedure. The first step in the procedure produces a tree like this:
```
<root>
  config:
    trainer: TrainerA
    model: ModelA
    dataset: DatasetA
    model_dataset@_global_: ModelA_DatasetA
```
(For simplicity I have omitted some details from the above, such as the `hydra/launcher` and `hydra/sweeper` settings)

This tree is then expanded recursively (in a depth-first manner), adding nodes from the defaults lists of the configs that appear in the tree. Of the config files appearing in the tree above (`TrainerA.yaml`, `ModelA.yaml`, `DatasetA.yaml`, and `ModelA_Dataset.yaml`), only the config `TrainerA` has a defaults list. Adding the defaults list from the file `TrainerA.yaml` to the tree results in the below:
```
<root>
  config:
    trainer: TrainerA
      trainer/criterion: CriterionB  # NOTE: CriterionB is used here due to command-line override.
    model: ModelA
    dataset: DatasetA
    model_dataset@_global_: ModelA_DatasetA
```
(The above is again simplified. For the real deal, run `python my_app.py --info defaults-tree model=ModelA dataset=DatasetA trainer/criterion=CriterionB`)

The defaults list in the file `TrainerA.yaml` has the default `criterion: CriterionA` in it's defaults list, but the defaults tree uses the default `criterion: CriterionB` instead because the override `trainer/criterion=CriterionB` was given at the command line.

At this point, the defaults tree is full-expanded; all of the relevant defaults lists from the input configs have been included in the tree.


### Flattening the defaults tree
The defaults tree is then flattened into a list of defaults (possibly with @package specifications). The flattening is done by traversing the tree depth-first, with child nodes coming before their parents. Here is the result of flattening:
```
* trainer/criterion/CriterionB
* trainer/TrainerA
* model/ModelA
* dataset/DatasetA
* model_dataset/ModelA_DatasetA @ _global_
* config
```
(Again, the above is simplified. To see the full list, run `python my_app.py --info defaults model=ModelA dataset=DatasetA trainer/criterion=CriterionB`)
As an example of "child nodes coming before their parents", `CriterionB` comes before its parent `TrainerA` in the list above. The last element in the list is `config` because `config` was is the most grandparental node in the defaults tree.

### Merging

Finally, to produce the output config, the input configs from the above list are merged (in order!). The file `ModelA_DatasetA.yaml` has higher priority than the file `CriterionB.yaml` because the former comes later in the above list. This explains why in your example we have `trainer.criterion.crit_param1==1337` instead of `trainer.criterion.crit_param1==888`.

## The Remedy

If you want `CriterionB` to take precedence over `ModelA_DatasetA`, then `CriterionB` must come later in the final list of configs to be merged. This can be achieved by putting the default `trainer: TrainerA` at the end of the defaults list in the file `config.yaml` (as mentioned in the TLDR section above).
```yaml
defaults:
  - model: ModelA
  - dataset: DatasetA
  - optional model_dataset@_global_: ${model}_${dataset}
  - trainer: TrainerA
```
This results in the following (expanded) defaults tree:
```
<root>
  config:
    model: ModelA
    dataset: DatasetA
    model_dataset@_global_: ModelA_DatasetA
    trainer: TrainerA
      trainer/criterion: CriterionB  # NOTE: CriterionB is used here due to command-line override.
```
Finally, the defaults tree is flattened as follows, resulting in CriterionB taking priority over ModelA_DatasetA.
```
* model/ModelA
* dataset/DatasetA
* model_dataset/ModelA_DatasetA @ _global_
* trainer/criterion/CriterionB
* trainer/TrainerA
* config
```

#### A note on interpolation in the defaults list
Hydra's current implementation requires that the interpolated default `model_dataset: ${model}_{dataset}` come later in the defaults list than the interpolation targets `model: ...` and `dataset: ...`.

**Comment by Jasha10** [comment]
The above can be parlayed into a Hydra docs page on "Recursive Defaults Lists in Detail".

**Comment by jbaczek** [comment]
Hi @Jasha10, thanks for swift and detailed answer! This brings a lot of insight!
But there is a behavior that I want to achieve and your proposed solution isn't exactly what I had in mind. The problem is like this:
When I use a combination of `ModelA` and `DatasetA` I want config to change it's default values so it looks exactly like this (note that `crit_param2` isn't present in the final config):
```
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetA
model:
  param: A
dataset:
  ds_param: aaa
trainer:
  criterion:
    crit_param1: 1337
  train_param: 13
```
But when I override it with command line override I want it to look like this:
```
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionB
model:
  param: A
dataset:
  ds_param: aaa
trainer:
  criterion:
    crit_param1: 888
    crit_param2: 999
  train_param: 13
```
And in other cases it should look like this:
```
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetB
model:
  param: A
dataset:
  ds_param: xxx
trainer:
  criterion:
    crit_param1: 111
    crit_param2: 222
  train_param: 13
```

The problem in your proposed solution is that `optional model_dataset@_global_: ${model}_${dataset}` doesn't take precedence over the default trainer so it results in this behavior:
```
root@jbaczek-ws:/workspace/tmp# python my_app.py -c job model=ModelA dataset=DatasetA
model:
  param: A
dataset:
  ds_param: aaa
trainer:
  criterion:
    crit_param1: 111
    crit_param2: 222
  train_param: 13
```
Is there currently a way to implement such behavior with hydra?

**Comment by Jasha10** [comment]
> Is there currently a way to implement such behavior with hydra?

My initial attempt was to create a file `trainer/criterion/CriterionAA.yaml`, which would contain the config `crit_param1: 1337`, and to add an element `override /trainer/criterion: CriterionAA` to the defaults list in the file `ModelA_DatasetA.yaml`. The motivation was that, given the specific combination ModelA+DatasetA, the default CriterionA would be overridden by the specialized CriterionAA.

This attempt did not work, as Hydra cannot handle the `override` keyword if it is used in the defaults list of an interpolated config; the defaults list of the file specified by interpolation `${model}_${dataset}` is not allowed to contain overrides.
```text
$ python my_app.py -c job model=ModelA dataset=DatasetA
model_dataset/ModelA_DatasetA: Default List Overrides are not allowed in the subtree
of an in interpolated config group (override trainer/criterion=CriterionAA).
```
This restriction is in place because defaults list interpolations are resolved after all overrides have been finalized. If this restriction were not in place, and if the interpolated config `ModelA_DatasetA.yaml` were to add new overrides to the override list, then the value of the interpolation `${model}_${dataset}` might change as a result of the new override.

In theory, I think it would be possible to allow an interpolated config (e.g. `ModelA_DatasetA.yaml`) use the "override" keyword in its defaults list (with well-defined behavior) so long as the new override does not affect whether ModelA_DatasetA (or any of it's ancestors) are to be included in the defaults tree. I feel that currently this would be very challenging to implement.

---

For now, the sure bet is to manually select the appropriate criterion using the command line:
```
# manually selecting criterion AA:
python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionAA
# manually selecting criterion B:
python my_app.py -c job model=ModelA dataset=DatasetA trainer/criterion=CriterionB
# Using the default criterion specified in TrainerA.yaml:
python my_app.py -c job model=ModelA dataset=DatasetB
```

**Comment by Jasha10** [comment]
I am closing this issue in favor of #1989.
